impeller/entity/contents/typed_contents_visitor.h

```diff
@@ -0,0 +1,61 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_IMPELLER_ENTITY_CONTENTS_TYPED_CONTENTS_VISITOR_H_
+#define FLUTTER_IMPELLER_ENTITY_CONTENTS_TYPED_CONTENTS_VISITOR_H_
+
+// All of these includes are unfortunate, but this is the price we pay to be
+// able to use std::is_base_of_v.  The alternative is to maintain that in code
+// which could introduce bugs.
+#include "impeller/entity/contents/anonymous_contents.h"
+#include "impeller/entity/contents/atlas_contents.h"
+#include "impeller/entity/contents/checkerboard_contents.h"
+#include "impeller/entity/contents/color_source_contents.h"
+#include "impeller/entity/contents/contents_visitor.h"
+#include "impeller/entity/contents/filters/filter_contents.h"
+#include "impeller/entity/contents/solid_rrect_blur_contents.h"
+#include "impeller/entity/contents/text_contents.h"
+#include "impeller/entity/contents/texture_contents.h"
+#include "impeller/entity/contents/vertices_contents.h"
+
+namespace impeller {
+
+template <typename BaseT, typename DerivedPtrT>
+struct IsDerivedOrSame
+    : std::integral_constant<bool,
+                             std::is_base_of_v<BaseT, DerivedPtrT> ||
+                                 std::is_same_v<BaseT, DerivedPtrT>> {};
+
+/// A ContentsVisitor that only visits Contents of type `T` and its subclasses.
+template <typename T>
+class TypedContentsVisitor : public ContentsVisitor {
```

Comment:
So this restricts the visit to ColorSource contents via a type check, but could we instread restrict the visition by giving only ColorSourceContents a visit method.

Consider:

```
class GeometryVisitor {
  void AddGeometry(...);
}


class Contents {
void Accept(GeometryVisitor* visitor) override {
  // no-op
}
}

class ColorSourceContents : public Contents {
  void Accept(GeometryVisitor* visitor) override {
    visitor->AddGeometry(GetGeometry());
  }
}
```

In fact, this is roughly equivalent (but more verbose) than adding a method onto contents to grab the geometry:

```
class Contents {
  std::optional<Geometry> GetGeometry() { return std::nullopt; }
}

class ColorSourceContents: public Geometry {
  std::optional<Geometry> GetGeometry() overrride;
```

So IMO, once you start stripping out the unnecessary visitors I think its more obvious that this should be a method.

Reply:
Yep, both of those things would technically work.  Those are both inferior designs though.  As you noted, the first one is basically equivalent to the second.  The second approach moves `Contents` in a direction where its interface is the union of all the possible functionality of the subclasses, where only a subset of the api is applicable depending on runtime information.  That makes the `Contents` abstraction difficult to understand.  It stops being a meaningful abstraction and just a result of consequence.

The first approach also suffers from not being an idiomatic implementation of the Visitor pattern.  It has the worst of both approaches of not having the flexibility of a proper Visitor and not being as straight forward as the virtual method on Contents.

The purpose of the Visitor pattern is to add the ability to expand the functionality of an abstraction without bloating the interface.  So, in this case we have one place in the code that wants the geometry of a Contents.  Why should we confuse the Contents abstraction by expanding its API for a need that only exists in one place in the code?  Having a proper visitor allows us to make those sorts of decisions about what should be graduated to the Contents API or not.  A dead giveaway that something shouldn't be in the Contents API is if it is only applicable to a certain subset of subclasses.

On `TypedContentsVisitor`, that's only a convenience.  I could remove it if you want.

I'm happy to talk about this more over gvc if you want.  I have a lot of experience designing systems.  I think we could be doing better and I'm happy to share.